apt-hook/hook.cc

```diff
@@ -201,69 +282,230 @@ int main(int argc, char *argv[])
+         if (res.disabled_esms_a > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d additional update is available with UA Apps: ESM.",
+                              "%d additional updates are available with UA Apps: ESM.",
+                              res.disabled_esms_a),
+                     res.disabled_esms_a);
+            ioprintf(std::cout, "\n");
+         }
+
+         ioprintf(std::cout, gettext("To see these additional updates run: apt list --upgradable"));
+         ioprintf(std::cout, "\n");
+         ioprintf(std::cout, gettext("See https://ubuntu.com/advantage or run: sudo ua status"));
          ioprintf(std::cout, "\n");
       }
-
-      ioprintf(std::cout, gettext("To see these additional updates run: apt list --upgradable"));
-      ioprintf(std::cout, "\n");
-      ioprintf(std::cout, gettext("See https://ubuntu.com/advantage or run: sudo ua status"));
-      ioprintf(std::cout, "\n");
+   } else if (subcommand == PreInvoke) {
+      if (command_used == "upgrade" || command_used == "dist-upgrade") {
+         // Print static message if present. Used for "Expiring soon" and "Expired but in grace period" messages.
+         std::ifstream message_file(CONTRACT_EXPIRY_STATUS_MESSAGE_PATH);
+         if (message_file.is_open()) {
+            std::cout << message_file.rdbuf();
+            message_file.close();
+         }
+
+         // Try not enabled messages
+         process_template_file(
+            ESM_APPS_NOT_ENABLED_MESSAGE_TEMPLATE_PATH,
+            ESM_APPS_NOT_ENABLED_MESSAGE_STATIC_PATH,
+            esm_a_packages_count,
+            space_separated_esm_a_packages,
+            esm_i_packages_count,
+            space_separated_esm_i_packages
+         );
+         process_template_file(
+            ESM_INFRA_NOT_ENABLED_MESSAGE_TEMPLATE_PATH,
+            ESM_INFRA_NOT_ENABLED_MESSAGE_STATIC_PATH,
+            esm_a_packages_count,
+            space_separated_esm_a_packages,
+            esm_i_packages_count,
+            space_separated_esm_i_packages
+         );
+
+         // Try command specific messages
+         if (command_used == "upgrade") {
+            process_template_file(
+               CONTRACT_EXPIRED_APT_UPGRADE_MESSAGE_TEMPLATE_PATH,
+               CONTRACT_EXPIRED_APT_UPGRADE_MESSAGE_STATIC_PATH,
+               esm_a_packages_count,
+               space_separated_esm_a_packages,
+               esm_i_packages_count,
+               space_separated_esm_i_packages
+            );
+         } else if (command_used == "dist-upgrade") {
+            process_template_file(
+               CONTRACT_EXPIRED_APT_DIST_UPGRADE_MESSAGE_TEMPLATE_PATH,
+               CONTRACT_EXPIRED_APT_DIST_UPGRADE_MESSAGE_STATIC_PATH,
+               esm_a_packages_count,
+               space_separated_esm_a_packages,
+               esm_i_packages_count,
+               space_separated_esm_i_packages
+            );
+         }
+
+         /*
+         // TODO remove
```

Comment:
+1 on removal of this, we can intrument the various message .tmpl files you need in /v/l/ubuntu-advantage/messages and make sure hook can be a simple string formatter that creates final message files.